File: navi/motor_ctl/motor.cpp
#include <iostream>
#include "motor.h"
#include "wiringPi.h"
#include "softPwm.h"

using namespace std;
using namespace MotorCTL;


Motor::Motor( int r_fpin, int r_bpin, int l_fpin, int l_bpin ) {
    wiringPiSetup();

    this->r_forward_pin = r_fpin;
    this->r_backward_pin = r_bpin;
    this->l_forward_pin = l_fpin;
    this->l_backward_pin = l_bpin;

    softPwmCreate( this->r_forward_pin, 0, 100 );
    softPwmCreate( this->r_backward_pin, 0, 100 );
    softPwmCreate( this->l_forward_pin, 0, 100 );
    softPwmCreate( this->l_backward_pin, 0, 100 );
}


void Motor::cleanup( void ) {
    pinMode( this->r_forward_pin, INPUT );
    pinMode( this->r_backward_pin, INPUT );
    pinMode( this->l_forward_pin, INPUT );
    pinMode( this->l_backward_pin, INPUT );
}


void Motor::move( Instruct instruction ) {
    if (instruction.dir) {
        // FORWARD or BACKWARD
        if (instruction.speed > 0) this->forward( instruction.speed );
        else this->backward();
    }
    else {
        // LEFT or RIGHT
        if (instruction.speed > 0) this->turn_left();
        else this->turn_right();
    }
    delay(5000);
}


void Motor::forward( int speed ) {
    cout << "FORWARD (" << speed << ")" << endl;
    softPwmWrite(this->r_forward_pin, 0);
    softPwmWrite(this->r_backward_pin, speed);
    softPwmWrite(this->l_forward_pin, 0);
    softPwmWrite(this->l_backward_pin, speed);
}


void Motor::backward( void ) {
    
}


void Motor::turn_left( void ) {
    
}


void Motor::turn_right( void ) {
    
}

File: functions/motor.py
from typing import *


class Motor:
    def __init__(self, pins: List[int]) -> None:
        self.r_fpin = pins[0]
        self.r_bpin = pins[1]
        self.l_fpin = pins[2]
        self.l_bpin = pins[3]

        '''
            Please add WiringPi codes, Here!
            Examples::
                >>> import wiringpi
                >>> ...
                >>> wiringpi.wiringPiSetup()
        '''

    def forward(self):
        '''
            Examples::
                >>> import wiringpi
                >>> wiringpi.softPwmWrite(pin, speed)
                >>> wiringpi.delay(delay)
        '''
        ...

    def backward(self):
        '''
            Turn right > Turn right > Forward
        '''
        ...

    def left(self):
        ...

    def right(self):
        ...

